### PR TITLE
AM-76-mfa-rate-limit-gateway

### PR DESCRIPTION
**Pull request by ashraf706**
This implementation of rate limit is applied for SMS and OTP factor. Current implementation is based on sliding token bucket algorithm.

### How to test

1. Uncomment `mfa_rate` section in gateway gravitee.yml and decrease the value for testing
2. Create SMS factor
3. In challenge page refresh the page so that it reaches `limit` value in gravitee.yml file
4. Now refresh again and you should get `mfa_request_limit_exceed` message.

### CODE REVIEW

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+public class MongoRateLimitRepository extends AbstractManagementMongoRepository implements RateLimitRepository {
+    private static final String RATE_LIMIT = "rate_limit";
+    private static final String FIELD_FACTOR_ID = "factorId";
+
+    private MongoCollection<RateLimitMongo> rateLimitCollection;
+
+    @PostConstruct
+    public void init() {
+        rateLimitCollection = mongoOperations.getCollection(RATE_LIMIT, RateLimitMongo.class);
+        super.init(rateLimitCollection);
```
Comment (leleueri) [inline gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java:53]: Index creation is missing.
Please name explicitly the indexes with a name less than 64 characters as there a limit on DocumentDB as far as I know.

File: gravitee-am-model/src/main/java/io/gravitee/am/model/RateLimit.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.Date;
+
+/**
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+
+public class RateLimit {
+    private String id;
+    private String userId;
```
Comment (leleueri) [inline gravitee-am-model/src/main/java/io/gravitee/am/model/RateLimit.java:27]: I think we should also add referenceId (initialized with the domainID) and referenceType (initialized to ReferenceType.DOMAIN) in order to be able to delete all the rate_limit by domain

**Review by leleueri — CHANGES_REQUESTED** [review]
In addition of the inner code comments, here is some feedbacks.

- we should remove all rate_limit entries linked to a user when this user is deleted
- we should remove all rate_limit entries linked to a domain on domain deletion.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java [hunk 3]
```diff
@@ -595,6 +619,10 @@ private void sendChallenge(FactorProvider factorProvider,
                 );
     }
 
+    private boolean shouldLimitRequest(Factor factor) {
+        return rateLimiterService.isRateLimitEnabled() && (FactorType.EMAIL.equals(factor.getFactorType()) || FactorType.SMS.equals(factor.getFactorType()));
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java]: changed as suggested

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java [hunk 4]
```diff
@@ -356,6 +383,10 @@ private void sendChallenge(FactorProvider factorProvider, RoutingContext routing
                 );
     }
 
+    private boolean shouldLimitRequest(Factor factor) {
+        return rateLimiterService.isRateLimitEnabled() && (FactorType.EMAIL.equals(factor.getFactorType()) || FactorType.SMS.equals(factor.getFactorType()));
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java]: changed as suggested

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/service/ratelimit/impl/RateLimiterServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class RateLimiterServiceImpl implements RateLimiterService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RateLimiterServiceImpl.class);
+
+    @Value("${mfa_rate.enabled:false}")
+    private boolean isEnabled;
+
+    @Value("${mfa_rate.limit:0}")
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/service/ratelimit/impl/RateLimiterServiceImpl.java:45]: still pending

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/service/ratelimit/impl/RateLimiterServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    public boolean isRateLimitEnabled() {
+        return isEnabled;
+    }
+
+    @Override
+    public Single<Boolean> tryConsume(String userId, String factorId, String client) {
+        if (timePeriod <= 0 || limit <= 0) {
+            return Single.just(false);
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/service/ratelimit/impl/RateLimiterServiceImpl.java:67]: warn log added

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java (hunk 1 above)
Comment (ashraf706) [reply gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java:54]: index added

File: gravitee-am-model/src/main/java/io/gravitee/am/model/RateLimit.java (hunk 2 above)
Comment (ashraf706) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/RateLimit.java:27]: referenceId and referenceType field added

**Comment by ashraf706** [comment]
> In addition of the inner code comments, here is some feedbacks.
> 
> * we should remove all rate_limit entries linked to a user when this user is deleted (when the user is deleted by the ManagementAPI but also through the SCIM handlet in the Gateway)
> * we should remove all rate_limit entries linked to a domain on domain deletion. (https://github.com/gravitee-io/gravitee-access-management/blob/master/gravitee-am-service/src/main/java/io/gravitee/am/service/impl/DomainServiceImpl.java#L487)

done

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Override
+    public Completable deleteByUser(String userId) {
+        return Completable.fromPublisher(rateLimitCollection.deleteMany(eq(FIELD_USER_ID, userId)));
+    }
+
+    @Override
+    public Completable deleteByDomain(String domainId) {
+        return Completable.fromPublisher(rateLimitCollection.deleteMany(eq(FIELD_REFERENCE_ID, domainId)));
```
Comment (leleueri) [inline gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java:103]: Maybe we can add here the ReferenceType.DOMAIN as second criteria.

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcRateLimitRepository.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        throw new RepositoryIllegalQueryException("Unable to delete from RateLimit with userId");
+    }
+
+    @Override
+    public Completable deleteByDomain(String domainId) {
+        LOGGER.debug("delete RateLimit with domain id '{}'", domainId);
+
+        Criteria whereClause = Criteria.empty();
+        if(domainId != null && !domainId.isEmpty()){
+            whereClause = whereClause.and(where("reference_id").is(domainId));
```
Comment (leleueri) [inline gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcRateLimitRepository.java:123]: Maybe we can add ReferenceType.DOMAIN as a second criteria

**Review by leleueri — APPROVED** [review]
LGTM, just minor comments

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java (hunk 7 above)
Comment (ashraf706) [reply gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/MongoRateLimitRepository.java]: done, added as suggested

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcRateLimitRepository.java (hunk 8 above)
Comment (ashraf706) [reply gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcRateLimitRepository.java]: done, added as suggested

**Pull request merged**